### DISCUSSION TITLE
Shorthand to exclude all files matching `outputs` glob from `inputs`

### DISCUSSION THREAD

**Discussion by reckter** [vercel/turborepo#7480] [Ideas]
### Which project is this feature idea for?

Turborepo

### Describe the feature you'd like to request

I think this sits somewhere between feature and bug.
Currently when you have an output directory that also matches an input glob, turbo treats all files in that output directory relevant for the cache-key. Effectively making every output-input combination it's own cache. Which intern hinders efficient caching.

Example: (other code left out for readability)

```json
{
    "pipeline": {
		"codegen": {
			"inputs": ["src/**"],
			"outputs": ["src/__generated__/**"]
		},
      }
}
```

### Describe the solution you'd like

Turbo should check if a file is matched by an output glob before considering it for the cache key.

### Describe alternatives you've considered

Negative matches would be a also a solution, but one the user has to find themselves.
Either with a negation-syntax in the "input" array, or a more explicit "ignore" field on tasks.

Workaround (that at least seams to work for now, but no definite answer yet was burned before by a wrong input-glob) is using negative character classes from https://github.com/bmatcuk/doublestar#character-classes :
``` inputs: ["src/[!_]*/**", "src/[!_]*"]```.
But you need to have 2 globs for that to work, and arguably it gets less readably this way.
And this probably doesn't work for all cases!

**Comment by nathanhammond** [comment]
One more-general note: typically build tools like Turborepo hand you a directory and _all_ of your outputs get dumped into that directory. It's distinct and disconnected from your repository. That you can colocate your generated assets with your source is both novel, and, as you can see, dangerous.

Normally we would tell you to ignore the output folder using `.gitignore`, but since you're manually specifying the input files, we no longer include that default behavior for you.

Here is the test case I used to reproduce your situation, on top of `create-turbo`:

```diff
diff --git a/apps/generated/package.json b/apps/generated/package.json
new file mode 100644
index 0000000..a376845
--- /dev/null
+++ b/apps/generated/package.json
@@ -0,0 +1,6 @@
+{
+  "name": "generated",
+  "scripts": {
+    "generate": "date >> src/output/date.txt"
+  }
+}
\ No newline at end of file
diff --git a/apps/generated/src/.gitignore b/apps/generated/src/.gitignore
new file mode 100644
index 0000000..9b1960e
--- /dev/null
+++ b/apps/generated/src/.gitignore
@@ -0,0 +1 @@
+output/
\ No newline at end of file
diff --git a/apps/generated/src/input.txt b/apps/generated/src/input.txt
new file mode 100644
index 0000000..81229b4
--- /dev/null
+++ b/apps/generated/src/input.txt
@@ -0,0 +1 @@
+date
\ No newline at end of file
diff --git a/turbo.json b/turbo.json
index b510892..151c04b 100644
--- a/turbo.json
+++ b/turbo.json
@@ -5,6 +5,10 @@
     "build": {
       "outputs": [".next/**"]
     },
+    "generate": {
+      "inputs": ["src/**"],
+      "outputs": ["src/output/**"]
+    },
     "lint": {},
     "dev": {
       "cache": false,
```

If you remove `inputs` from `turbo.json` you get the behavior you want because our default behavior of respecting `.gitignore`.

The _ideal_ pattern for your case is actually this:

```
    "codegen": {
      "inputs": ["src/**", "!src/__generated__/**"],
      "outputs": ["src/__generated__/**"]
    },
```

We actually _can't_ make the change as you're proposing it because any task can write to its inputs (or the inputs of other tasks) and that isn't something that we can say is or is not valid from our position outside of the task.

The only option we have from our perspective would be to do something like a microsyntax, for example `"inputs": ["src/**", "!$outputs"]` to enable opt in.

I'm not going to close this, but I am going to rename the issue, and hold it open to see who else shows up to discuss and gauge community interest.

**Comment by reckter** [comment]
The output of this task is actually on a .gitignore allready, hovewer using the default `inputs` value would also mean including more files into the cache-key, as are necessary. (And that's kinda the point why I looked into this in the first place)

It might be fine, but still having the input be as tight as possible is our endgoal here.

Allowing true-negative globs, would probably be the go to solution then. 
(Edit see post below)
~~Eventhough I don't agree, that a task itself should be able to influence it's own input (inputs of other task for sure!), but that might just be a opinionated look into "task" in general.
A task that modifies its output is not really stable, in the sense that you need to run it until you reach a fix-point.
That kinda goes against my definition of a cachable-task in the first place.~~

But that doesn't mean that turbo, shouldn't be able to work with such task, so:
negative globs would work well for us! Tried those in the past, but quickly realized, that they don't work with the current implementation (At least, that's how I read the code, past issues here, and your "ideal  pattern" statemenet :) ). Please correct me if I am wrong here!

having a microsyntax like you proposed is a handy shortcut, but definitely needs good documentation for discoverability, it might even make sense to expose it more by making it a field on tasks? `ignoreOutputs: true`.
That way changing the default might be a thing turbo could consider.

**Comment by reckter** [comment]
Sometimes in life you write something, and then go do a thing, that just gives you the exact opposite insight as you just wrote:
Linting task with autofixing of course write their own input!.
Which also means they need to be run twice to be cachable (first time fixes the listing errors, second time writes the cache).

And especially for these linting task you don't want the output ignored!
So ignore the whole paragraph about my opinion above. It needs to be configurable :)

**Comment by nathanhammond** [comment]
To be clear, the "ideal" sketch in my previous comment here works in the current version, you don't need to go all the way into doublestar glob implementation details.

!-prefixed patterns are negated; much simpler to write than the complicated pattern you came up with.

I personally agree that outputs shouldn't clobber inputs, but the system isn't designed that way right now and we can't make that change without making it opt in. The ongoing tragedy of having something that people are using where we can't break people to fix our own past transgressions.

!gitignore could also be a way to specify include excludes. Or maybe !gitindex because you can have files in index that are otherwise ignored.

**Comment by nathanhammond** [comment]
Also, making it a flag whose default state we can toggle at a major version bump is a clever proposal! That's easier to help people through the situation.

**Comment by nathanhammond** [comment]
(Yup, lint --fix is our canonical example. And it reaches fixed point in two.)

**Comment by reckter** [comment]
The !-prefixed patterns didn't work in my testing (and trying it again, still don't in v1.8.3).
I can see that they will work in 1.8.4 (https://github.com/vercel/turbo/releases/tag/v1.8.4-canary.4) though!
So that's nice!
We will need to change it to that then, as it's way more readable than my workaround for sure :)

"!gitignore" or "!gitindex" (or maybe even a more special looking solution like `!$gitignore`) would also be nice. And would also be a solution for this case, even though I prefer `"!src/__generated__/**"`, as it requires less knowledge of the setup to be understood.

**Comment by anthonyshew** [comment]
I've got some thoughts on this one:
1. Microsyntaxes are not a zero cost abstraction. Any microsyntax that we add means there's a new, special set of knowledge that Turborepo developers would have to know and understand. The value that the microsyntax provides has to be **vastly** larger than the cost of the abstraction. (I don't think that would be the case here, as you'll see.)
2. Turborepo going out of its way to not do what you say doesn't seem like a smart thing to do, in my opinion. If I told Turborepo "these are the inputs and these are the outputs", I would want it to believe me. Following the principle of least surprise, I think I'd be pretty confused to find out that I can affect `inputs` by some value that I added to `outputs`. I'd prefer Turborepo do exactly what I told it to do.
3. I'm seeing some discussion about how using `inputs` drops all the defaults for what `turbo` usually infers as inputs. Between the date this proposal was created and today, we shipped a `$TURBO_DEFAULTS$` microsyntax for inputs that allows you to use the usual defaults but add on your adjustments for your specific `inputs`. (This is a great example of a microsyntax that adds a ton of value beyond its cost.)
4. This code block is clear to me as someone who is seeing it for the first time.
```		
"inputs": ["src/**"],
"outputs": ["src/__generated__/**"]
```
(4 cont) These globs overlap each other and, if I want to exclude the outputs, I need to add a negative glob. Additionally, if that negative glob were to be added, the intent would still be clear.
```		
"inputs": ["src/**", "!src/__generated__/**"],
"outputs": ["src/__generated__/**"]
```
5. This sort of problem can be fixed pretty quickly but not placing generated code into source code. I would say that, in the general sense, I would advise someone who is doing this to make sure there codegen isn't dumping into the same place where they are writing code (unless we're talking about code gen that is meant to be user code like v0 or GitHub Copilot).

At this point, you can probably tell that I'm not necessarily in favor of this one. 😄 This use case seems to have solutions that are fairly clear both by editing `inputs` to express what you really want it to be or by not generating code into user code. Adding some extra magic to save what is likely a user error seems to be a bridge too far.

**Comment by BrodBlox09** [comment]
> Microsyntaxes are not a zero cost abstraction. Any microsyntax that we add means there's a new, special set of knowledge that Turborepo developers would have to know and understand. The value that the microsyntax provides has to be vastly larger than the cost of the abstraction. (I don't think that would be the case here, as you'll see.)

A microsyntax like `$OUTPUTS$` that refers to the contents of `"outputs"`, similar to `$TURBO_DEFAULT$` which refers to the default `"inputs"`, abstracts minimally, as it simply refers directly to a visible value that is likely a few lines away from where the user is currently editing their `turbo.json` file. Imagine a complicated `"outputs"` entry for a non-trivial task; having to rewrite that verbatim but negated, while cumbersome but easily done for the initial setup, means that if `"outputs"` is changed at all, `"inputs"` is now out-of-sync and the user needs to remember that they need to update `"inputs"` when `"outputs"` is updated or their configuration file will now silently be telling turbo to no longer create the proper cache keys (albeit, this would technically be the user's fault). Adding a variable/microsyntax that simply expands to reference the value of `"outputs"` for `"inputs"` would be not only a nice QoL addition (especially if `!$OUTPUTS$` negates the value of `"outputs"` for you), but also make it less likely for maintainers to mess up when updating `turbo.json`.

In my case, I have my packages set up in the following manner:
```
package-root/
- dist/
- src/
- <package-specific directories>/
```
`<package-specific directories>/` represents arbitrary directories used for each package, which might be, e.g., reusable templates (`templates/`) or other resources for that specific package that don't fit in `src` but still serve as inputs for a `build` task. In my `turbo.json` configuration file, it would be clearer for me to write the following:
```
"inputs": [ "$TURBO_DEFAULT$", "!$OUTPUTS$" ],
"outputs": [ "dist/**" ]
```
Instead of:
```
"inputs": [ "$TURBO_DEFAULT$", "!dist/**" ],
"outputs": [ "dist/**" ]
```
